ratis-netty/src/main/java/org/apache/ratis/netty/server/NettyServerStreamRpc.java

```diff
@@ -115,10 +136,33 @@ private ChannelInboundHandler getServerHandler(){
       public void channelRead(ChannelHandlerContext ctx, Object msg) {
         final DataStreamRequestByteBuf request = (DataStreamRequestByteBuf)msg;
         final ByteBuf buf = request.slice();
-        final AtomicBoolean released = new AtomicBoolean();
-        streams.computeIfAbsent(request.getStreamId(), id -> getDataStreamFuture(buf, released))
-            .thenApply(stream -> writeTo(buf, stream, released.get()))
-            .thenAccept(byteWritten -> sendReply(request, byteWritten, ctx));
+        final AtomicBoolean shouldRelease = new AtomicBoolean();
+
+        CompletableFuture<?>[] parallelWrites = new CompletableFuture<?>[clients.size() + 1];
+        CompletableFuture<?> localWrites =
+            streams.computeIfAbsent(request.getStreamId(), id -> getDataStreamFuture(buf, shouldRelease))
+            .thenApply(stream -> writeTo(buf, stream, shouldRelease.get()));
+        parallelWrites[0] = localWrites;
+        peersStreamOutput.putIfAbsent(request.getStreamId(), getDataStreamOutput());
+
+        for (int i = 0; i < clients.size(); i++) {
+          // do not need to forward header request
+          if (request.getStreamOffset() == -1) {
+            DataStreamClientImpl.DataStreamOutputImpl impl =
+                (DataStreamClientImpl.DataStreamOutputImpl) peersStreamOutput.get(request.getStreamId()).get(i);
+            parallelWrites[i + 1] = impl.getHeaderFuture();
+          } else {
+            CompletableFuture<?> remoteWrites;
+            if (shouldRelease.get()) {
+              remoteWrites = new CompletableFuture<>();
+            } else {
+              remoteWrites =
+                  peersStreamOutput.get(request.getStreamId()).get(i).writeAsync(request.slice().nioBuffer());
+            }
+            parallelWrites[i + 1] = remoteWrites;
+          }
+        }
```

Comment:
Similar, use isHeader for the remote writes
```
        if (isHeader) {
          // header
          final List<DataStreamOutput> outs = getDataStreamOutput();
          peersStreamOutput.put(request.getStreamId(), outs);
          for (int i = 0; i < outs.size(); i++) {
            parallelWrites[i] = outs.get(i).getHeaderFuture();
          }
        } else {
          // body
          final List<DataStreamOutput> outs =  peersStreamOutput.get(request.getStreamId());
          for (int i = 0; i < clients.size(); i++) {
            parallelWrites[i] = outs.get(i).writeAsync(request.slice().nioBuffer());
          }
        }
```

Also, add getHeaderFuture() to DataStreamClient.